### PR TITLE
Add a relation spec for the database backup manager

### PR DESCRIPTION
**Pull request by shayancanonical**
[jira ticket](https://warthogs.atlassian.net/browse/DPE-910)

# Description
We (Data Platform) are in the near future planning to implement a database backup manager. A Juju admin will then be allowed to interact with this database backup manager to trigger a backup or a restore of a database via an action. Eventually, the backup manager will also contain a mechanism to schedule backups.

This PR outlines the first draft of the relation spec between the database backup manager and its related databases. It is expected that the database backup manager will be related to either the S3 Integrator (under development) or S3 Proxy charm (to be spec'ed and implemented), from which it will be able to get access necessary to S3 parameters.

### CODE REVIEW

File: interfaces/database_backup_manager/v0/README.md [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+The following is the criteria that a Provider and Requirer need to adhere to be compatible with this interface.
+
+### Provider
+
+- Is expected to detect a change in `job-id` and run either a backup or restore based on the `job-type`.
+- Is expected to only run a job when it is idle. Any job further requests after a backup/restore is started may be ignored.
+- Is expected to have a locking mechanism to avoid running multiple jobs at once.
+- Is expected to indicate to the Requirer the status of the current or most recently completed job via the `status` field.
+- Is expected to upload the log file for the job to the `s3-path`.
+- Is expected to upload the backup file to the `s3-path` if it was successful in backing up the database.
```
Comment (jnsgruk) [inline interfaces/database_backup_manager/v0/README.md:33]: Should this be `s3-backup-path` and the other be `s3-restore-path` just to make things more explicit?

File: interfaces/database_backup_manager/v0/README.md [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+- Is expected to retrieve the backup file from the `s3-backup-path` if it was required to restore a backup.
```
Comment (jnsgruk) [inline interfaces/database_backup_manager/v0/README.md:34]: As above - `s3-restore-path`?

File: interfaces/database_backup_manager/v0/README.md (hunk 1 above)
Comment (jnsgruk) [reply interfaces/database_backup_manager/v0/README.md:33]: Ah hang on, looks like the log file path is inconsistent here with the actual spec...

**Review by jnsgruk — APPROVED** [review]
Looks like there are some inconsistencies in the wording up top, but overall I think this is looking good.

File: interfaces/database_backup_manager/v0/README.md (hunk 1 above)
Comment (shayancanonical) [reply interfaces/database_backup_manager/v0/README.md]: I really like the idea of a separate backup and restore file path. I've also refactored the `storage interface data` and `storage interface type` into separate fields (suggested in offline feedback) in be75dd215a5cfb50f81f77eeae125cad0a84023b.

File: interfaces/database_backup_manager/v0/README.md (hunk 2 above)
Comment (shayancanonical) [reply interfaces/database_backup_manager/v0/README.md]: Addressed in be75dd215a5cfb50f81f77eeae125cad0a84023b

**Comment by shayancanonical** [comment]
Refactored `storage-interface` and `<storage_interface>_config` (`s3_config`) as described above in 3b4ad2a216547bd154b32e0d9a87ec47167ea7b8. Also, further fleshed out the JSON schemas for the requirer and provider in the same commit

File: interfaces/database_backup_manager/v0/README.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+1. restore the state of a related data from a specified backup in storage
+
+Initially, the backup or restore will be triggered manually against the database backup manager via an action. However, eventually, the juju admin will be able to schedule periodic backups on the database backup manager.
+
+## Direction
+
+```mermaid
+flowchart TD
+    Requirer -- job-id, job-type, storage-interface, storage-config, artifact-path --> Provider
+    Provider -- job-id, job-type, status, artifact-path --> Requirer
```
Comment (delgod) [inline interfaces/database_backup_manager/v0/README.md:17]: The response should store all the needed info for restoration.
```suggestion
    Provider -- job-id, job-type, status, artifact-path, storage-interface, storage-config --> Requirer
```

File: interfaces/database_backup_manager/v0/README.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+1. backup the state of a related database to a specified storage
+1. restore the state of a related data from a specified backup in storage
+
+Initially, the backup or restore will be triggered manually against the database backup manager via an action. However, eventually, the juju admin will be able to schedule periodic backups on the database backup manager.
+
+## Direction
+
+```mermaid
+flowchart TD
+    Requirer -- job-id, job-type, storage-interface, storage-config, artifact-path --> Provider
```
Comment (delgod) [inline interfaces/database_backup_manager/v0/README.md:16]: I am not sure what is the correct spelling in British English. we need to follow British English.
maybe
```suggestion
    Requirer -- job-id, job-type, storage-interface, storage-config, artefact-path --> Provider
```

File: interfaces/database_backup_manager/v0/README.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+1. backup the state of a related database to a specified storage
+1. restore the state of a related data from a specified backup in storage
+
+Initially, the backup or restore will be triggered manually against the database backup manager via an action. However, eventually, the juju admin will be able to schedule periodic backups on the database backup manager.
+
+## Direction
+
+```mermaid
+flowchart TD
```
Comment (delgod) [inline interfaces/database_backup_manager/v0/README.md:15]: Maybe we need to add also `job-config`, which can contain only `compression: bool` for now.
`job-config` should be passed in both directions

File: interfaces/database_backup_manager/v0/README.md [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+    related-units:
+      mysql/0:
+        in-scope: true
+        data:
+          jobs: '
+            [
+              {
+                "job-id": "another-unique-job-id",
+                "job-type": "restore",
+                "status": "error"
```
[the remark this answers is not in the document]
Comment (jnsgruk) [reply interfaces/database_backup_manager/v0/README.md:64]: I wouldn't recommend concatenating fields like that, we probably want to be more prescriptive, like:

```javascript
{
  ...,
  "status": "error",    // this should be limited to a few like success|error
  "message": "MySQL charm in blocked state",
  ...
}

**Review by gruyaume — CHANGES_REQUESTED** [review]

File: interfaces/database_backup_manager/v0/README.md [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+
+## Direction
+
+```mermaid
+flowchart TD
+    Requirer -- job-id, job-type, storage-interface, storage-config, artifact-path --> Provider
+    Provider -- job-id, job-type, status, artifact-path --> Requirer
+```
+
+The interface consists of two parties: a Provider (database charm) and a Requirer (database manager charm). The Requirer will be expected to provide the relevant jobs (backups or restores) to execute. The provider will then be expected to execute these jobs and respond with their statuses.
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply interfaces/database_backup_manager/v0/README.md]: addressed in e9efe849fae6727a19321e3a9837043daec08e3e

File: interfaces/database_backup_manager/v0/README.md [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+## Behavior
+
+The following is the criteria that a Provider and Requirer need to adhere to be compatible with this interface.
+
+### Provider
+
+- Is expected to detect any new jobs and run either a backup or restore based on the `job-type`.
+- Is expected to run concurrent backups, but only one restore at a time.
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply interfaces/database_backup_manager/v0/README.md]: addressed in e9efe849fae6727a19321e3a9837043daec08e3e

File: interfaces/database_backup_manager/v0/README.md [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+- Is expected to relay the results of completed jobs back to the Juju admin.
+- Is expected to preserve the history of completed and running jobs, but also provide a mechanism to clear the history of completed jobs.
+
+## Relation Data
+
+### Provider
+
+[\[JSON Schema\]](./schemas/provider.json)
+
+Provider (database charm) is expected to record the status of running or completed jobs in the **unit** databag of the unit that performed the job. In the case of a backup, the information will be placed in the **unit** databag of the unit that performed the backup. In the case of a restore, the information will be placed in the **unit** databag of the lowest numbered unit (where the restore is performed).
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply interfaces/database_backup_manager/v0/README.md]: addressed in e9efe849fae6727a19321e3a9837043daec08e3e

File: interfaces/database_backup_manager/v0/README.md (hunk 3 above)
Comment (shayancanonical) [reply interfaces/database_backup_manager/v0/README.md]: addressed in e9efe849fae6727a19321e3a9837043daec08e3e

File: interfaces/database_backup_manager/v0/README.md (hunk 4 above)
Comment (shayancanonical) [reply interfaces/database_backup_manager/v0/README.md]: addressed in e9efe849fae6727a19321e3a9837043daec08e3e

File: interfaces/database_backup_manager/v0/README.md (hunk 5 above)
Comment (shayancanonical) [reply interfaces/database_backup_manager/v0/README.md]: i also think it's a good idea to have `job-config`. addressed in e9efe849fae6727a19321e3a9837043daec08e3e

File: interfaces/database_backup_manager/v0/README.md [hunk 10]
```diff
@@ -0,0 +1,121 @@
+# `database_backup_manager`
```
[the remark this answers is not in the document]
Comment (shayancanonical) [reply interfaces/database_backup_manager/v0/README.md]: good point - addressed in e9efe849fae6727a19321e3a9837043daec08e3e

**Review by gruyaume — APPROVED** [review]

**Review by simskij — APPROVED** [review]

**Pull request merged**